File: native-wrapper/android-native/GnhBackgroundSync/GnhBackgroundSyncModule.kt
package im.getnowhere.app.backgroundsync

import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule

/** RN bridge: register WebView injector + receive background-sync responses. */
class GnhBackgroundSyncModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "GnhBackgroundSync"

    @ReactMethod
    fun registerWebViewInjector() {
        RemoteNodeSyncBridgeHolder.setScriptInjector { script ->
            if (reactApplicationContext.hasActiveReactInstance()) {
                reactApplicationContext
                    .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
                    .emit("gnhBackgroundSyncInject", script)
            }
        }
    }

    @ReactMethod
    fun clearWebViewInjector() {
        RemoteNodeSyncBridgeHolder.setScriptInjector(null)
    }

    @ReactMethod
    fun resolveBackgroundSync(requestId: String, outcome: String) {
        RemoteNodeSyncBridgeHolder.resolveRequest(
            requestId,
            RemoteNodeSyncBridgeHolder.parseOutcome(outcome),
        )
    }
}

File: native-wrapper/ios-native/GnhBackgroundSync/GnhBackgroundSyncModule.swift
import Foundation
import React

@objc(GnhBackgroundSync)
class GnhBackgroundSyncModule: RCTEventEmitter {
    private var hasListeners = false

    override static func requiresMainQueueSetup() -> Bool { true }

    override func supportedEvents() -> [String]! {
        ["gnhBackgroundSyncInject"]
    }

    override func startObserving() {
        hasListeners = true
    }

    override func stopObserving() {
        hasListeners = false
    }

    @objc func registerWebViewInjector() {
        RemoteNodeSyncBridgeHolder.shared.setScriptInjector { [weak self] script in
            guard let self, self.hasListeners else { return }
            self.sendEvent(withName: "gnhBackgroundSyncInject", body: script)
        }
    }

    @objc func clearWebViewInjector() {
        RemoteNodeSyncBridgeHolder.shared.setScriptInjector(nil)
    }

    @objc func resolveBackgroundSync(_ requestId: String, outcome: String) {
        RemoteNodeSyncBridgeHolder.shared.resolveRequest(
            requestId: requestId,
            outcome: RemoteNodeSyncOutcome(outcomeString: outcome),
        )
    }
}
